src/skeleton/frontend.ts

```diff
@@ -165,41 +181,62 @@ class RenderHelper extends RefCounted {
           this.defineCommonShader(builder);
           this.defineAttributeAccess(builder);
           defineLineShader(builder);
+          defineNoOpNodeSetters(builder);
           builder.addAttribute("highp uvec2", "aVertexIndex");
           builder.addUniform("highp float", "uLineWidth");
+          builder.addUniform("highp float", "uNodeDiameter");
+          builder.addVarying("vec4", "vColor");
+          builder.addVertexCode(glsl_COLORMAPS);
+          builder.addVertexCode(glsl_nanometersToPixels);
+          builder.addVertexCode(`
+float nanometersToPixels(float value) {
+  vec2 lineOffset = getLineOffset();
+  highp vec3 vertexA = readAttribute0(aVertexIndex.x);
+  highp vec3 vertexB = readAttribute0(aVertexIndex.y);
+  highp vec3 vertex = mix(vertexA, vertexB, lineOffset.x);
+  return nanometersToPixels(value, vertex, uProjection, uViewModel, uLineParams.xy);
+}
+float ng_lineWidth;
+void setLineWidth(float width) {
+  ng_lineWidth = width;
+}
+void emitRGB(vec3 color) {
+  vColor = vec4(color.rgb, uColor.a);
+}
+void emitRGBA(vec4 color) {
+  vColor = color;
+}
+`);
           let vertexMain = `
 highp vec3 vertexA = readAttribute0(aVertexIndex.x);
 highp vec3 vertexB = readAttribute0(aVertexIndex.y);
-emitLine(uProjection, vertexA, vertexB, uLineWidth);
 highp uint lineEndpointIndex = getLineEndpointIndex();
 highp uint vertexIndex = aVertexIndex.x * lineEndpointIndex + aVertexIndex.y * (1u - lineEndpointIndex);
+highp uint vertexIndex2 = aVertexIndex.y * lineEndpointIndex + aVertexIndex.x * (1u - lineEndpointIndex);
```

Comment:
I am a bit confused why I needed to add this other vertexIndex. The original seems to access the data for the opposite node for each line segment.